File: fortran/wired.f90
! Wired Benchmarks - Fortran Implementation
! (c) 2024 ryan@freestatelabs


module wired
    use omp_lib

    implicit none 
    public cross, crosscols, crosscols2, subcols, multmv, cross2
    public createproblem, createproblem2
    public solve, solvecm, solve2, solve3

    contains

    ! Make the test problem with row-major ordering
    subroutine createproblem(Nn, Ns, nodes, sources)

        integer :: Nn, Ns, i
        real, dimension(:,:) :: nodes, sources 

        real :: zstart, zend, dz, z0, z1

        zstart = -1000.0 
        zend = 1000.0 
        dz = (zend - zstart)/Ns 
        z0 = zstart 
        z1 = dz 
        
        DO i = 1,Ns
            sources(1,i) = 0.0
            sources(2,i) = 0.0
            sources(3,i) = z0
            sources(4,i) = 0.0
            sources(5,i) = 0.0
            sources(6,i) = z1
            sources(7,i) = 5000.0
            z0 = z1 
            z1 = z1 + dz 
        END DO 
        DO i = 1,Nn
            nodes(1,i) = 3.0
            nodes(2,i) = 3.0
            nodes(3,i) = 0.0
        END DO 

    end subroutine createproblem 


    ! Create the problem with column-major ordering
    subroutine createproblem2(Nn, Ns, nodes, sources)

        integer :: Nn, Ns, i
        real, dimension(:,:) :: nodes, sources 

        real :: zstart, zend, dz, z0, z1

        zstart = -1000.0 
        zend = 1000.0 
        dz = (zend - zstart)/Ns 
        z0 = zstart 
        z1 = dz 
        
        DO i = 1,Ns
            sources(1,i) = 0.0
            sources(2,i) = 0.0
            sources(3,i) = z0
            sources(4,i) = 0.0
            sources(5,i) = 0.0
            sources(6,i) = z1
            sources(7,i) = 5000.0
            z0 = z1 
            z1 = z1 + dz 
        END DO 
        DO i = 1,Nn
            nodes(i,1) = 3.0
            nodes(i,2) = 3.0
            nodes(i,3) = 0.0
        END DO 

    end subroutine createproblem2 


    ! Compute c = a x b for 3-element vectors
    pure function cross(a, b) result(c)

        real, dimension(3) :: c
        real, dimension(3), intent(in) :: a, b
        
        c(1) = a(2) * b(3) - a(3) * b(2)
        c(2) = a(3) * b(1) - a(1) * b(3)
        c(3) = a(1) * b(2) - a(2) * b(1)

    end function cross


    ! Compute c = a x b for 3xN matrices, across the columns of a using vector b
    subroutine crosscols(c, a, b) 

        real, dimension(:,:) :: c
        real, dimension(:,:), intent(in) :: a 
        real, dimension(:), intent(in) :: b
        
        c(1,:) = a(2,:) * b(3) - a(3,:) * b(2)
        c(2,:) = a(3,:) * b(1) - a(1,:) * b(3)
        c(3,:) = a(1,:) * b(2) - a(2,:) * b(1)

    end subroutine crosscols


    ! Compute c = a x b for Nx3 matrices, across the rows of a using vector b
    subroutine crosscols2(c, a, b) 

        real, dimension(:,:) :: c
        real, dimension(:,:), intent(in) :: a 
        real, dimension(:), intent(in) :: b
        
        c(:,1) = a(:,2) * b(3) - a(:,3) * b(2)
        c(:,2) = a(:,3) * b(1) - a(:,1) * b(3)
        c(:,3) = a(:,1) * b(2) - a(:,2) * b(1)

    end subroutine crosscols2


    ! Subtract columns of b from vector a
    subroutine subcols(c, a, b) 

        real, dimension(:,:) :: c
        real, dimension(:), intent(in) :: a 
        real, dimension(:,:), intent(in) :: b
        integer i

        c(:,1) = -b(:,1) + a(1) 
        c(:,2) = -b(:,2) + a(2) 
        c(:,3) = -b(:,3) + a(3)

    end subroutine subcols


    ! Compute c = a x b by modifying vector c ("in-place") in an attempt to 
    ! reduce memory allocations.
    ! Does not appear to be any faster.
    subroutine cross2(c, a, b)

        real, dimension(3) :: c
        real, dimension(3), intent(in) :: a, b
        
        c(1) = a(2) * b(3) - a(3) * b(2)
        c(2) = a(3) * b(1) - a(1) * b(3)
        c(3) = a(1) * b(2) - a(2) * b(1)

    end subroutine cross2


    ! Take the norm of the columns of A and place them in b
    subroutine normcols(b, A) 

        real, dimension(:) :: b 
        real, dimension(:,:) :: A 

        b(:) = (A(:,1)**2 + A(:,2)**2 + A(:,3)**2)**0.5
    end subroutine normcols 


    ! Take the dot product of the columns of A with vector b and place them in c 
    subroutine dotcols(c, A, b) 

        real, dimension(:) :: c, b 
        real, dimension(:,:) :: A

        c(:) = A(:,1)*b(1) + A(:,2)*b(2) + A(:,3)*b(3) 

    end subroutine dotcols


    ! Multiply Matrix A row-wise by vector b and place in C
    subroutine multmv(C, A, b) 

        real, dimension(:,:) :: C, A 
        real, dimension(:) :: b 

        C(:,1) = A(:,1) * b(:) 
        C(:,2) = A(:,2) * b(:)
        C(:,3) = A(:,3) * b(:)

    end subroutine multmv


    ! Solve the Biot-Savart problem with opt-level 1: simple loops 
    ! This is the row-major method (slower)
    function solve(nodes, sources) result (bfield)

        real, dimension(:,:), intent(in) :: nodes, sources 
        real, dimension(:,:), allocatable :: bfield

        ! Instantiate intermediate values 
        real, dimension(3) :: a, b, c, cxa 
        real :: mu0, d
        integer :: Nn, Ns, i, j

        allocate(bfield(3,ubound(nodes,2)))
        bfield = 0.0    ! initialize all elements to zero
        mu0 = 1e-7

        Nn = size(nodes, 2) 
        Ns = size(sources, 2)

        do j = 1,Ns 
            d = mu0 * sources(7,j)

            do i = 1,Nn 
                a = sources(4:6,j) - sources(1:3,j) 
                b = sources(1:3,j) - nodes(:,i)
                c = sources(4:6,j) - nodes(:,i) 
                cxa = cross(c,a)
                cxa = cxa * d * (norm2(cxa)**(-2)) * (dot_product(a,c)/norm2(c) - dot_product(a,b)/norm2(b))
                bfield(:,i) = bfield(:,i) + cxa             
            end do  

        end do 

    end function solve 


    ! Solve the Biot-Savart problem with opt-level 1: simple loops in column-major
    ! ordering (faster)
    function solvecm(nodes, sources) result (bfield)

        real, dimension(:,:), intent(in) :: nodes, sources 
        real, dimension(:,:), allocatable :: bfield

        ! Instantiate intermediate values 
        real, dimension(3) :: a, b, c, cxa 
        real :: mu0, d
        integer :: Nn, Ns, i, j

        allocate(bfield(ubound(nodes,1),3))
        bfield = 0.0 
        mu0 = 1e-7

        Nn = size(nodes, 1) 
        Ns = size(sources, 2)

        do j = 1,Ns 
            d = mu0 * sources(7,j)

            do i = 1,Nn 
                a = sources(4:6,j) - sources(1:3,j) 
                b = sources(1:3,j) - nodes(i,:)
                c = sources(4:6,j) - nodes(i,:) 
                cxa = cross(c,a)
                cxa = cxa * d * (norm2(cxa)**(-2)) * (dot_product(a,c)/norm2(c) - dot_product(a,b)/norm2(b))
                bfield(i,:) = bfield(i,:) + cxa             
            end do  

        end do 

    end function solvecm


    ! Solve the biot-savart problem with opt-level 2: vectorization
    function solve2(nodes, sources) result (bfield)

        real, dimension(:,:), intent(in) :: nodes, sources 
        real, dimension(:,:), allocatable :: bfield

        ! Intermediate values 
        real, dimension(3) :: a
        real, dimension(:,:), allocatable :: b, c, cxa
        real, dimension(:), allocatable :: norm_cxa, dot_ac, norm_c, dot_ab, norm_b 
        real :: mu0, d
        integer :: Nn, Ns, i, j, thread_id

        Ns = size(sources,2) 
        Nn = size(nodes,1)

        allocate(bfield(Nn,3))
        allocate(b(Nn,3))
        allocate(c(Nn,3))
        allocate(cxa(Nn,3))
        allocate(norm_cxa(Nn))
        allocate(dot_ac(Nn))
        allocate(norm_c(Nn))
        allocate(dot_ab(Nn))
        allocate(norm_b(Nn))
        bfield = 0.0    ! initialize entire array to zero

        mu0 = 1e-7

        do j = 1,Ns 
            d = mu0 * sources(7,j)
            a = sources(4:6,j) - sources(1:3,j) 

            ! Subtract matrics from vector row by row 
            call subcols(b, sources(1:3,j), nodes)
            call subcols(c, sources(4:6,j), nodes)

            call crosscols2(cxa, c, a)
            call normcols(norm_cxa, cxa) 
            call dotcols(dot_ac, c, a) 
            call dotcols(dot_ab, b, a) 
            call normcols(norm_b, b) 
            call normcols(norm_c, c)
            call multmv(cxa,  cxa, d * (norm_cxa(:)**(-2)) * (dot_ac(:)/norm_c(:) - dot_ab(:)/norm_b(:)))
            
            bfield(:,:) = bfield(:,:) + cxa(:,:)             
        end do  

    end function solve2


    ! Partition a matrix with length `N` into smaller segments that can be processed
    ! on `Nt` number of threads. `it` is the current thread number. Returns 
    ! a 2-length vector corresponding to start/stop indices
    function partition(it, Nt, N) result(i) 

        integer, intent(in) :: it, Nt, N 
        integer :: rem, Nperthread
        integer,dimension(2) :: i

        Nperthread = N / Nt  ! Integer division returns only quotient 
        rem = mod(N, Nt)

        if (it == 1) then 
            i(1) = 1 
            i(2) = i(1) + Nperthread - 1
            ! print*,"it = ",it," spans ", i(1), i(2)
        else if (it == Nt) then 
            i(2) = N 
            i(1) = i(2) - Nperthread - rem + 1 
            ! print*,"it = ",it," spans ", i(1), i(2)
        else
            i(1) = (it-1)*Nperthread + 1
            i(2) = i(1) + Nperthread - 1 
            ! print*,"it = ",it," spans ", i(1), i(2)
        end if 

    end function partition 


    ! Solve using parallel processing via OpenMP
    function solve3(nodes, sources, Nt) result(bfield) 

        real, dimension(:,:), intent(in) :: nodes, sources 
        integer :: Nt
        real, dimension(:,:), allocatable :: bfield, bfield_partial
        integer :: Ns, Nn, it, thread_id 
        integer, dimension(2) :: ispan

        Ns = ubound(sources,2)
        Nn = ubound(nodes,1)

        allocate(bfield(Nn,3))
        allocate(bfield_partial(Nn,3))
        bfield = 0.0
        bfield_partial = 0.0

        if (Nt == 0) then
            Nt = OMP_GET_NUM_THREADS()
        end if 
        print*,"Solving with ", Nt, " threads."

        !$OMP parallel private(bfield_partial) shared(bfield)

        !$OMP do
        do it = 1,Nt
            ispan = partition(it, Nt, Ns) 
            bfield_partial = solve2(nodes, sources(:,ispan(1):ispan(2)))
        end do
        !$OMP end do

        !$OMP critical 
        bfield = bfield + bfield_partial 
        !$OMP end critical 

        !$OMP end parallel

    end function solve3 

end module wired

File: c/wired.c
// Wired.jl benchmarks - C implementation 
// (c) 2024 ryan@freestatelabs

#include <stdio.h>
#include <omp.h>
#include <math.h>
#include <stdlib.h>

// Convenience function to create a 2D array of zeros
double** zeros(int Nrows, int Ncols) {

    double** array; 
    array = calloc(Nrows, sizeof(double*));
    for (int i=0; i<Nrows; i++) {
        array[i] = calloc(Ncols, sizeof(double));
    }

    return array;
}


// Create an array of pointer arrays for nodes
double** createnodes(int Nn) {

    int M = 3;
    double** nodes = zeros(M, Nn);

    for (int i=0; i<Nn; i++) {
        nodes[0][i] = 3.0;
        nodes[1][i] = 3.0; 
        nodes[2][i] = 0.0;
    }

    return nodes;
}


// De-allocate a 2D array
void free2darray(double** array, int Ncols) {

    for (int i=0; i<Ncols; i++) {
        free(array[i]);
    }

    free(array);
}


// Norm of a 3xN matrix along direction N
void normcols(double* b, double** A, int N) {
    for (int i=0; i<N; i++) {
        b[i] = sqrt(pow(A[0][i],2) +pow(A[1][i],2) +pow(A[2][i],2));
    }
}


// Cross the columns of A with vector b
void crosscols(double** C, double** A, double* b, int N) {

    for (int i=0; i<N; i++) {
        C[0][i] = A[1][i] * b[2] - A[2][i] * b[1];
        C[1][i] = A[2][i] * b[0] - A[0][i] * b[2];
        C[2][i] = A[0][i] * b[1] - A[1][i] * b[0];
    }
}


// Return an array of pointer arrays for the sources
double** createsources(int Ns) {

    int M = 7; 
    double** sources = zeros(Ns, M);

    double zstart = -1000.0;
    double zend = 1000.0;
    double zstep = (zend-zstart)/Ns;
    double z = zstart;
    double Ic = 5000.0;

    for (int i=0; i<Ns; i++){
        sources[i][0] = 0.0;
        sources[i][1] = 0.0;
        sources[i][2] = z; 
        sources[i][3] = 0.0;
        sources[i][4] = 0.0;
        sources[i][5] = z + zstep; 
        sources[i][6] = Ic;
        z += zstep; 
    }

    return sources;
}


// Solve with optimization level 2: auto-vectorization using the compiler
double** solve2(double** nodes, double** sources, int Nn, int Ns) {

    double** bfield = zeros(3, Nn);

    double* a = calloc(Nn, sizeof(double)); 
    double** b = zeros(3,Nn);
    double** c = zeros(3,Nn);
    double** cxa = zeros(3,Nn);
    double* norm_cxa = calloc(Nn, sizeof(double)); 
    double* dot_ac = calloc(Nn, sizeof(double)); 
    double* dot_ab = calloc(Nn, sizeof(double)); 
    double* norm_c = calloc(Nn, sizeof(double)); 
    double* norm_b = calloc(Nn, sizeof(double)); 
    double* s = calloc(Nn, sizeof(double)); 
    double d;

    // Outer loop over sources
    for (int j=0; j<Ns; j++) {

        d = (1e-7) * sources[j][6];
        a[0] = sources[j][3] - sources[j][0];
        a[1] = sources[j][4] - sources[j][1];
        a[2] = sources[j][5] - sources[j][2];

        // Inner loop over nodes
        for (int i=0; i<Nn; i++) {
            b[0][i] = sources[j][0] - nodes[0][i];
            b[1][i] = sources[j][1] - nodes[1][i];
            b[2][i] = sources[j][2] - nodes[2][i];

            c[0][i] = sources[j][3] - nodes[0][i];
            c[1][i] = sources[j][4] - nodes[1][i];
            c[2][i] = sources[j][5] - nodes[2][i];
        }

        crosscols(cxa, c, a, Nn);
        normcols(norm_cxa, cxa, Nn);
        normcols(norm_b, b, Nn);
        normcols(norm_c, c, Nn);

        for (int i=0; i<Nn; i++) {
            dot_ac[i] = a[0]*c[0][i] + a[1]*c[1][i] + a[2]*c[2][i];
        }
            
        for (int i=0; i<Nn; i++) {
            dot_ab[i] = a[0]*b[0][i] + a[1]*b[1][i] + a[2]*b[2][i];
        }

        for (int i=0; i<Nn; i++) {
            s[i] = d * (dot_ac[i]/norm_c[i] - dot_ab[i]/norm_b[i]) * pow(norm_cxa[i],-2);
        }

        for (int i=0; i<Nn; i++) {
            bfield[0][i] += cxa[0][i] * s[i]; 
            bfield[1][i] += cxa[1][i] * s[i]; 
            bfield[2][i] += cxa[2][i] * s[i];
        }
    }    

    free(a); 
    free2darray(b, 3);
    free2darray(c, 3);
    free2darray(cxa, 3);
    free(norm_cxa);
    free(dot_ac);
    free(dot_ab);
    free(norm_b);
    free(norm_c);
    free(s);

    return bfield;
}


// Chunk a matrix of major length `N` for processing on `Nt` threads
//  `it` is the number of the current thread
// Modifies a 2-length array `i` of the indices for processing
void partition(int i[], int it, int Nt, int N){

    int rem = N % Nt; 
    int Nperthread = (N - rem) / Nt; 

    if (it == 0)  {
        i[0] = 0;
        i[1] = i[0] + Nperthread;
    }
    else if (it == Nt-1) {
        i[1] = N;
        i[0]= i[1] - Nperthread - rem;
    }
    else { 
        i[0] = it*Nperthread;
        i[1] = i[0] + Nperthread;
    }
}
         

// Solve with optimization level 3: parallel processing
double** solve3(double** nodes, double** sources, int Nn, int Ns) {

    double source[7];
    double** bfield = zeros(3, Nn);
    double** bfield_partial = zeros(3,Nn);
    double** sources_partial = zeros(Ns,7);
    int slice[2] = {0, 0};

    int i, j, k, it, row, Nt;
    #pragma omp parallel 
    {
        Nt = omp_get_num_threads();
    }
    int length_partition = 0;

    #pragma omp parallel private(i, j, k, it, row, slice, length_partition, bfield_partial, sources_partial) shared(bfield, sources) 
    {
        #pragma omp for ordered 
        for (it=0; it<Nt; it++) {
                partition(slice, it, Nt, Ns);
            }
    
        #pragma omp for
        for (it=0; it<Nt; it++) {
            partition(slice, it, Nt, Ns);
            length_partition = slice[1] - slice[0];

            // Create the partioned matrix 
            double** sources_partial = zeros(length_partition, 7);

            row = 0;
            for (j=slice[0]; j<slice[1]; j++) { 

                for (k=0; k<7; k++) {
                    sources_partial[row][k] = sources[j][k];
                }
                row++;
            }

            bfield_partial = solve2(nodes, sources_partial, Nn, length_partition); 
            free2darray(sources_partial, 7);

            #pragma omp critical 
            {
                for (i=0; i<3; i++) {
                    for (j=0; j<Nn; j++) {
                        bfield[i][j] += bfield_partial[i][j];
                    }
                }
            }
        }
    }
    
    return bfield;
}
